src/libtools/graph.cc

```diff
@@ -0,0 +1,129 @@
+/*
+ *            Copyright 2009-2018 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <algorithm>
+#include <string>
+#include <votca/tools/graph.h>
+
+using namespace std;
+
+namespace votca {
+namespace tools {
+
+class GraphNode;
+
+void Graph::updateIds_(Graph& g) {
+  if (!this->id_set_) {
+    this->calcId_();
+    this->id_set_ = true;
+  }
+  if (!g.id_set_) {
+    g.calcId_();
+    g.id_set_ = true;
+  }
+}
+
+bool Graph::operator!=(Graph& g) {
```

Comment:
I would never expect that `compare` operator will update the objects..

Reply:
It is not really updating the object in the sense that it is changing what's in the graph, it is only creating an id that summarizes the contents of the graph in a string. It is easier to compare a string than to compare all the contents each time I want to see if two graphs are equal. The id itself is private so the user is not supposed to know or use it.

Reply:
Actually, if you look into the docs (https://en.cppreference.com/w/cpp/language/operator_comparison) this kind of operator should be `const`. I am simply worried about the unexpected side effect.

Reply:
Oh I see, I think I can make it constant. If the id is always up to date, then there is no reason to call that update function when the equivalence operator is called. Good point!

Reply:
Wow you just helped me find a design flaw in this class :) thanks.